packages/api/endpoints/granules.js

```diff
@@ -140,37 +141,50 @@ async function del(req, res) {
   log.info(`granules.del ${granuleId}`);
 
   const granuleModelClient = new models.Granule();
+  const granulePgModel = new GranulePgModel();
+  const filePgModel = new FilePgModel();
 
-  let granule;
+  const knex = await getKnexClient({ env: process.env });
+
+  let dynamoGranule;
+  let pgGranule;
   try {
-    granule = await granuleModelClient.getRecord({ granuleId });
+    dynamoGranule = await granuleModelClient.getRecord({ granuleId });
+    pgGranule = await granulePgModel.get(knex, { granule_id: granuleId });
   } catch (error) {
     if (error instanceof RecordDoesNotExist) {
       return res.boom.notFound(error);
     }
     throw error;
   }
 
-  if (granule.detail) {
-    return res.boom.badRequest(granule);
+  // TODO what is this?
+  if (dynamoGranule.detail) {
+    return res.boom.badRequest(dynamoGranule);
   }
 
-  try {
-    await granuleModelClient.delete(granule);
-  } catch (error) {
-    if (error instanceof DeletePublishedGranule) {
-      return res.boom.badRequest(error.message);
-    }
-    throw error;
-  }
+  await knex.transaction(async (trx) => {
+    // Delete the granule's files from Postgres and S3
+    await filePgModel.deleteGranuleFiles(trx, pgGranule);
+    // Delete the Granule from Postgres
+    await granulePgModel.delete(trx, pgGranule);
+    // Delete the granule from Dynamo
+    await granuleModelClient.delete(dynamoGranule);
+  })
+    .catch((error) => {
+      if (error instanceof DeletePublishedGranule) {
+        return res.boom.badRequest(error.message);
+      }
+      throw error;
+    });
```

Comment:
This is the main flow:

1 - Delete files from Postgres and S3
2 - Delete Granule from Postgres
3 - Delete Granule from Dynamo (and try to delete files from S3 again. Could remove this but we have a test that indicates that its fine to try and not delete anything from S3. Note below).

This is a little different from the original plan in that here we're deleting files before granules. The reason being a File has a foreign key pointing to a Granule. We cannot delete a granule while a File still exists that refers to it in this way.

There is a potential situation here where, say your Granule Postgres delete fails. In that situation we will have already deleted the files from S3. We can roll back everything else but the S3 files are gone.

I could split out the S3 file deletion so it doesn't happen in the same step as the PG file deletion but I opted not to do it this way because then the primary function to delete files on the File model (File.deleteGranuleFiles) will remove them from PG but not S3. I think we want them coupled so they're (hopefully) always removed from both places.

Open to discussion here. There are lots of options.

Reply:
Something to keep in mind is that delete from the PG model doesn't actually throw an error if it fails to delete.
https://github.com/knex/knex/blob/master/lib/query/querybuilder.js#L1144-L1151

Reply:
Thanks @markdboyd and @botanical! So after the architecture discussion yesterday this is the new flow I'll be trying:

```
Transaction 1
  Delete Files
    Transaction 2
      Delete each file from PG
      Delete each file from S3
  Delete Granule from PG
  Delete Granule from Dynamo
```
If Transaction 2 fails, we'll fail Transaction 1.

How does that sound?

@botanical I'll definitely add a test around failed PG deletes to ensure they trigger the rollback as we'd expect. I think right now the tests we have around this transaction trigger a failed state by throwing errors before the actual PG delete. We can see what happens with the actual delete fails.

Reply:
This is addressed over a series of PRs but the new logic is here:

https://github.com/nasa/cumulus/blob/19d1a6cb4175af68b8e1ddc1d7c27a30ea3ebdc7/packages/api/endpoints/granules.js#L170

And here's the actual delete function:

https://github.com/nasa/cumulus/blob/19d1a6cb4175af68b8e1ddc1d7c27a30ea3ebdc7/packages/api/lib/granules.js#L81